tests/utils/test_price_utils.py

```diff
@@ -0,0 +1,577 @@
+                test_case["normalized_days_remaining"]
+            )
+
+            np.testing.assert_almost_equal(
+                apr,
+                test_case["expected_result"],
+                err_msg="unexpected apr"
+            )
+
+
+    def test_calc_spot_price_from_apr(self):
+        """Unit tests for the calc_spot_price_from_apr function"""
+
+        test_cases = [
+            # test 1: 10% apr; 6mo remaining;
+            {
+                "apr_decimal": 0.10, # 10% apr
+                "normalized_days_remaining": 0.5, # 6 months = 0.5 years
+                # price = 1 / (1 + 0.10 * 0.5)
+                #     = 0.1052631579
+                "expected_result": 0.9523809524 # just over 0.95
+            },
+            # test 2: 2% apr; 6mo remaining;
+            {
+                "apr_decimal": 0.02, # 2% apr
+                "normalized_days_remaining": 0.5, # 6 months = 0.5 years
+                # price = 1 / (1 + 0.02 * 0.5)
+                #     = 0.9900990099
+                "expected_result": 0.9900990099 # just over 0.99
+            },
+            # test 3: 0% apr; 6mo remaining;
+            {
+                "apr_decimal": 0, # 0% apr
+                "normalized_days_remaining": 0.5, # 6 months = 0.5 years
+                # price = 1 / (1 + 0 * 0.5)
+                #     = 1
+                "expected_result": 1
+            },
+            # test 4: 21% apr; 3mo remaining;
+            {
+                "apr_decimal": 0.21, # 21% apr
+                "normalized_days_remaining": 0.25, # 3 months = 0.25 years
+                # price = 1 / (1 + 0.21 * 0.25)
+                #     = 0.2105263158
+                "expected_result": 0.9501187648 # just over 0.95
+            },
+            # test 5: 5% apr; 12mo remaining;
+            {
+                "apr_decimal": 0.05, # 5% apr
+                "normalized_days_remaining": 1, # 12 months = 1 year
+                # price = 1 / (1 + 0.05 * 1)
+                #     = 0.05263157895
+                "expected_result": 0.9523809524 # just over 0.95
+            },
+            # test 6: 0.10 price; 3mo remaining;
+            {
+                "apr_decimal": 36, # 3600% apr
+                "normalized_days_remaining": 0.25, # 3 months = 0.25 years
+                # price = 1 / (1 + 0 * 0.5)
+                #     = 0
+                "expected_result": 0.10
+            }
+        ]
+
+        for test_case in test_cases:
+            spot_price = price_utils.calc_spot_price_from_apr(
+                test_case["apr_decimal"],
+                test_case["normalized_days_remaining"]
+            )
+
+            np.testing.assert_almost_equal(
+                spot_price,
+                test_case["expected_result"],
+                err_msg="unexpected apr"
+            )
+
+
+    # ### YieldSpace ###
+
+
+    def test_calc_k_const(self):
+        """Unit tests for calc_k_const function"""
+
+        test_cases = [
```

Comment:
i like before how you had test cases for bond and share reserves == 0, add those here too!